sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/tuplesketchesAggregates.scala

```diff
@@ -0,0 +1,843 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.expressions.aggregate
+
+import org.apache.datasketches.tuple.{Intersection, Sketch, Summary, Union, UpdatableSketch, UpdatableSketchBuilder, UpdatableSummary}
+
+import org.apache.spark.SparkUnsupportedOperationException
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.expressions.{ExpectsInputTypes, Expression, ExpressionDescription, Literal}
+import org.apache.spark.sql.catalyst.expressions.aggregate.TypedImperativeAggregate
+import org.apache.spark.sql.catalyst.trees.{QuaternaryLike, TernaryLike}
+import org.apache.spark.sql.catalyst.util.{ArrayData, CollationFactory, ThetaSketchUtils}
+import org.apache.spark.sql.errors.QueryExecutionErrors
+import org.apache.spark.sql.internal.types.StringTypeWithCollation
+import org.apache.spark.sql.types.{AbstractDataType, ArrayType, BinaryType, DataType, DoubleType, FloatType, IntegerType, LongType, StringType, StructType}
+import org.apache.spark.unsafe.types.UTF8String
+
+sealed trait TupleSketchState {
+  def serialize(): Array[Byte]
+  def eval(): Array[Byte]
+}
+case class UpdatableTupleSketchBuffer[U, S <: UpdatableSummary[U]](sketch: UpdatableSketch[U, S])
+    extends TupleSketchState {
+  override def serialize(): Array[Byte] = sketch.compact.toByteArray
+  override def eval(): Array[Byte] = sketch.compact.toByteArray
+}
+case class UnionTupleAggregationBuffer[S <: Summary](union: Union[S]) extends TupleSketchState {
+  override def serialize(): Array[Byte] = union.getResult.toByteArray
+  override def eval(): Array[Byte] = union.getResult.toByteArray
+}
+case class IntersectionTupleAggregationBuffer[S <: Summary](intersection: Intersection[S])
+    extends TupleSketchState {
+  override def serialize(): Array[Byte] = intersection.getResult.toByteArray
+  override def eval(): Array[Byte] = intersection.getResult.toByteArray
+}
+case class FinalizedTupleSketch[S <: Summary](sketch: Sketch[S]) extends TupleSketchState {
+  override def serialize(): Array[Byte] = sketch.toByteArray
+  override def eval(): Array[Byte] = sketch.toByteArray
+}
+
+/**
+ * The TupleSketchAgg function utilizes a Datasketches TupleSketch instance to count a
+ * probabilistic approximation of the number of unique values in a given column with associated
+ * summary values, and outputs the binary representation of the TupleSketch.
+ *
+ * See [[https://datasketches.apache.org/docs/Tuple/TupleSketches.html]] for more information.
+ *
+ * @param child
+ *   child expression (struct with key and summary value) against which unique counting will occur
+ * @param lgNomEntriesExpr
+ *   the log-base-2 of nomEntries decides the number of buckets for the sketch
+ * @param summaryType
+ *   the type of summary (double, integer, string)
```

Comment:
I am curious here as to why we are deciding to accept the data type of the input value as an explicit argument. With the KLL quantiles functions, we encoded this type into the function names themselves, e.g. 

```
kll_sketch_agg_bigint(col, k)
kll_sketch_agg_float(col, k)
kll_sketch_agg_double(col, k)
```

Applying the same idea here, we'd have:

```
tuple_sketch_agg_bigint
tuple_sketch_agg_float
tuple_sketch_agg_double
```

This would bring additional consistency and prevent the need for parsing and processing the additional argument.

Reply:
Yes that's fair. I thought a bit about this and the reason I went with the single function was to limit the number of SQL functions created which will cause a good amount of redundancy. If we break it down per summary type, this will will result in ~33 functions instead of 11 (tuple has set operations with other tuple and theta sketches leads to an explosion). That said, I agree it would improve type safety and align better with Spark SQL conventions. I’m open to refactoring toward typed variants if we think that’s the right direction! One challenge would be functions like `tuple_sketch_estimate`, breaking that function down by summary type would be a bit awkward. We’d need to decide whether to encode the summary type in the function name or infer it from the sketch input.

Reply:
Added sql named params to all functions that take both lgNomEntries and mode